aten/src/ATen/native/TopKImpl.h

```diff
@@ -20,69 +20,124 @@ void topk_impl_loop(
     TensorAccessor<const scalar_t, 1> tmp_values(
         reinterpret_cast<scalar_t*>(data[2] + i * strides[2]),
-        &dim_size, &tmp_values_stride);
-
-    auto n_2 = dim_size;
-    auto use_partial_sort = k * 64 <= n_2;
+        &dim_size,
+        &tmp_values_stride);
 
-    for (const auto j : c10::irange(n_2)) {
+    for (const auto j : c10::irange(use_partial_sort ? k : dim_size)) {
       queue[j].first = tmp_values[j];
       queue[j].second = j;
     }
 
     // we want nan to be sorted as top for numpy compatibility
     if (use_partial_sort) {
+      auto queue_begin = queue.begin();
+      auto queue_end = queue.begin() + k;
+      auto& queue_front_ref = queue.front();
+      auto& queue_back_ref = queue.back();
       if (largest) {
-        std::partial_sort(queue.begin(), queue.begin() + k, queue.end(),
-          [](const elem_t& x, const elem_t& y) -> bool {
-            return ((_isnan<accscalar_t>(x.first) && !_isnan<accscalar_t>(y.first)) || (x.first > y.first));
-          });
+        auto comp = [](elem_t const& a, elem_t const& b) {
+          return (_isnan(a.first) && !_isnan(b.first)) || (a.first > b.first);
+        };
+        std::make_heap(queue_begin, queue_end, comp);
+
+        for (auto idx = k; idx < dim_size; ++idx) {
+          elem_t this_val{tmp_values[idx], idx};
+          if (comp(this_val, queue_front_ref)) {
+            std::pop_heap(queue_begin, queue_end, comp);
+            queue_back_ref = this_val;
+            std::push_heap(queue_begin, queue_end, comp);
+          }
+        }
+
+        if (sorted) {
+          std::sort_heap(queue_begin, queue_end, comp);
+        }
       } else {
-        std::partial_sort(queue.begin(), queue.begin() + k, queue.end(),
-          [](const elem_t& x, const elem_t& y) -> bool {
-            return ((!_isnan<accscalar_t>(x.first) && _isnan<accscalar_t>(y.first)) || (x.first < y.first));
-          });
+        auto comp = [](elem_t const& a, elem_t const& b) {
+          return (!_isnan(a.first) && _isnan(b.first)) || (a.first < b.first);
+        };
+        std::make_heap(queue_begin, queue_end, comp);
+
+        for (auto idx = k; idx < dim_size; ++idx) {
+          elem_t this_val{tmp_values[idx], idx};
+          if (comp(this_val, queue_front_ref)) {
+            std::pop_heap(queue_begin, queue_end, comp);
+            queue_back_ref = this_val;
+            std::push_heap(queue_begin, queue_end, comp);
+          }
+        }
+
+        if (sorted) {
+          std::sort_heap(queue_begin, queue_end, comp);
+        }
       }
     } else {
       if (largest) {
-        std::nth_element(queue.begin(), queue.begin() + k - 1, queue.end(),
-          [](const elem_t& x, const elem_t& y) -> bool {
-            return ((_isnan<accscalar_t>(x.first) && !_isnan<accscalar_t>(y.first)) || (x.first > y.first));
-          });
-        if (sorted) {
-          std::sort(queue.begin(), queue.begin() + k - 1,
+        std::nth_element(
```

Comment:
Are all of these modifications only formatting changes? If so, it might be desirable to submit a separate, small PR in which we make the formatting changes and then make the substantiative changes in this PR - that'll make the change surface smaller if a revert is necessary.